src/core/public/http/fetch.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Url from 'url';
+
+import { merge } from '../../utils';
+import { IBasePath, HttpInterceptor, HttpHandler, HttpFetchOptions, IHttpResponse } from './types';
+import { HttpFetchError } from './http_fetch_error';
+import { HttpInterceptController } from './http_intercept_controller';
+import { HttpResponse } from './response';
+import { interceptRequest, interceptResponse } from './intercept';
+import { HttpInterceptHaltError } from './http_intercept_halt_error';
+
+interface Params {
+  basePath: IBasePath;
+  kibanaVersion: string;
+}
+
+const JSON_CONTENT = /^(application\/(json|x-javascript)|text\/(x-)?javascript|x-json)(;.*)?$/;
+const NDJSON_CONTENT = /^(application\/ndjson)(;.*)?$/;
+
+export class FetchService {
+  private readonly interceptors = new Set<HttpInterceptor>();
+
+  constructor(private readonly params: Params) {}
+
+  public intercept(interceptor: HttpInterceptor) {
+    this.interceptors.add(interceptor);
+    return () => this.interceptors.delete(interceptor);
+  }
+
+  public removeAllInterceptors() {
```

Comment:
Do we still need it? I suspect it's used in tests only. I'd prefer not to make it a part of the public API. Not a blocker for the current PR.

Reply:
Correct, it's only used by tests at this time. I'd like to move the detailed interceptor tests out of the overall HttpService tests and into a test that only exercises this class. At that point, we can remove this completely from the public API.

However, I wanted to make sure not to change the tests too much in this initial refactor so as to avoid regressions.

Reply:
> However, I wanted to make sure not to change the tests too much in this initial refactor so as to avoid regressions.

💯